core/client/fs/src/main/java/alluxio/client/block/stream/SharedGrpcDataReader.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.stream;
+
+import alluxio.client.file.FileSystemContext;
+import alluxio.grpc.ReadRequest;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.NioDataBuffer;
+import alluxio.resource.LockResource;
+import alluxio.wire.WorkerNetAddress;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A shared gRPC data reader that cache blocks data for multi-thread accessing.
+ *
+ * It follows GrpcDataReader protocol and takes strong assumption:
+ * Parallel read to the same file happens on the same time, so that read request is
+ * serialized by kernel
+ */
+@NotThreadSafe
+public class SharedGrpcDataReader implements DataReader {
+  private static final Logger LOG = LoggerFactory.getLogger(SharedGrpcDataReader.class);
+  private static final int BLOCK_LOCK_NUM = 32;
+  private static final ReentrantReadWriteLock[] BLOCK_LOCKS;
+  /** A map from block id to the block's cached data reader. */
+  @GuardedBy("BLOCK_LOCKS")
+  private static final ConcurrentHashMap<Long, BufferCachingGrpcDataReader> BLOCK_READERS
+      = new ConcurrentHashMap<>();
+
+  static {
+    BLOCK_LOCKS = new ReentrantReadWriteLock[BLOCK_LOCK_NUM];
+    for (int i = 0; i < BLOCK_LOCK_NUM; i++) {
+      BLOCK_LOCKS[i] = new ReentrantReadWriteLock();
+    }
+  }
+
+  private static ReentrantReadWriteLock getBlockLock(long blockId) {
+    return BLOCK_LOCKS[(int) (blockId % BLOCK_LOCKS.length)];
```

Comment:
There is no "hashing" happening to the block id. block ids are a concatenation of 2 numbers: `<container id>:<sequence number>`. For all files, the first block of the file will always start with sequence number 0. That means the first block of every file may always hash to the the same block lock, the second block of all files will be the next block, and so on. There would be a pattern.

Is behavior ok? I would imagine that we would want a more "random" distribution of the block locks?

Reply:
good catch! Added a hash func to distribute the lock